### DISCUSSION TITLE
Attaching entities for the notify platform

### DISCUSSION THREAD

**Discussion by AngellusMortis** [home-assistant/architecture#705] [Entity Models]
## Background

There is a need to have a way to sign URLs for external use so they can securely be access anonymously (i.e. for push notifications for companion apps). The way the `camera` platform currently does this is by adding an `entity_picture` attribute to the entity with a rotating access token. This is bad because it is adding an access token to the state machine.

Some other useful places for this are for motion or objection detection (as above). If you have a camera capable of doing motion detection and generates real time event thumbnails, attaching those to notifications is rather useful so we pattern used for `camera` can be avoided and not proliferated. 

## Suggestion

Implement a method on entities that allows you to resolve an image URL. Then add options to the notify platform to "attach" an entity. Attaching the entity / method from the entity should allow for creating a signed URL for anonymous access.

### Ideas

* Method: something like `async_get_image(external: bool = False)`, `external` signs and constructs external URL automatically
* Maybe limit method to only camera, binary sensor, sensor and media player entities? Is limiting necessary?
* How do we "select" entities that support this? Maybe a new selector/entity attribute?
* For notify services: add an entity option that lets you filter for entities with `async_get_image` support and a second option that lets you sign the URL and construct an external URL (maybe default that to true?)
* After this is implemented, we can deprecate and start the process of removing URLs as attributes (like `camera.entity_picture`)
* Maybe a way to generate these URLs in a template? Not sure if that is needed if it is baked into the notify service.

### Potential Issues

* The current signing process requires a Websocket connection, which in turns requires a user. Automations do not run with users. 
    * Do we have a "secret key" somewhere stored that we can use to sign tokens for automations for?

## References 

* Related PR that spawned discussion: https://github.com/home-assistant/core/pull/63696
* How we currently sign URLs for media players: https://github.com/home-assistant/core/blob/dev/homeassistant/components/media_source/__init__.py#L122-L140

**Comment by balloob** [comment]
So I think that there are 2 proposals in 1 issue here, they are related but not the same.

1. Allow sending an entity to the notify platform and let platforms resolve it. Ie, allow attaching `camera.kitchen` in a notification call and let the `notify` integration, or it's platforms, resolve this into a snapshot or feed.
1. Some notification integrations send URLs instead of the content of files. In the case of a camera snapshot, Home Assistant hosts this and the receiver needs to be able to fetch this data without having authentication.

I think that both are good ideas. We might want to split the topics though.

**Reply by AngellusMortis** [reply]
They definitely can be separate, but I do not think the second one necessarily needs to occur here. `camera_proxy` already exists and it is entirely possible we do not need to host the images in HA itself (for the example of UniFi Protect, we can proxy the request from the NVR).

I think the first step is attaching the entity to the notification. The second step would potentially allows us to make a generic view to service said content from that integration can use when generating a URL.

I think especially in the case of a NVR for a security camera system, we should avoid directly serving the files ourselves if we can. My NVR has over 50TB of storage. My HA machine only has 256GB.

EDIT: Just to clarify, this discussion I only intended to be for 1. 2 can be a follow up one. This would assume you already _have_ a URL to sign.

**Reply by balloob** [reply]
Alright let's focus on 1. 

Current `notify` services support `message`, `title`, `target` and `data` (catch-all) ([source](https://github.com/home-assistant/core/blob/dev/homeassistant/components/notify/services.yaml#L6)).

I suggest we add a new field to attach things.

```yaml
service: notify.balloob
data:
  message: Someone at the door!
  attach:
    - entity: camera.front_door
    - media_source: "media-source://local/bla"
```

The notify integration can provide helpers to integrations to help resolve attachments to streams and/or snapshots. Snapshots could resolve to bytes or URL

**Reply by AngellusMortis** [reply]
Would we want the notify integration to handle resolving the content for the entity as well? Or would want to entity class to handle it? My thought each entity would handle their own notification attachment to make it easier to for integrations to extend in the future.

**Reply by balloob** [reply]
I would expect the notify integration to provide a helper method and let platforms call it to resolve it. The notify integration helper would then use helpers of the `camera` and `media_source` integrations to resolve the actual data.

**Reply by zacwest** [reply]
I think the different attachments should support different representations too, at least at the underlying notify level. For example, iOS can handle image, video and audio attachments; Android can only handle image. We wouldn't want to handicap one of the notify targets because another doesn't handle it.

**Reply by zacwest** [reply]
Another behavior of the iOS for cameras is that it'll stream the MJPEG/HLS versions of the entity, which are more of a 'live' view than retrieving an existing image/content. It might be worth considering this when building out the representations of an attachment, e.g. a camera could be:

- mp4 snapshot
- jpeg snapshot
- hls stream
- mjpeg stream

(or some subset therein)

**Reply by AngellusMortis** [reply]
For sure. To start, I think it may be good to start with just images though. We can iterate over to add more types. I think mp4 snapshot and HLS would be good next targets since MP4 files are not much different an image (fixed length) and HLS stream are already supported by the `stream` integration.

The `data` attributes could still be used for the media types not yet supported.

**Comment by zacwest** [comment]
Some of the notify services (for example, iOS/Android) go in plain text over third party servers. (At least until I wrap up home-assistant/iOS#1679 and home-assistant/core#52048 for iOS.) Although I think the trustworthiness of Apple and Google is relatively high, it might be worth allowing a particular notify service to opt out of the presigned URLs. I can see the value in avoiding a roundtrip request to refresh the access token if necessary, though -- but that may be cleaner to solve by extending the lifetime of the access token to a longer window.

**Reply by AngellusMortis** [reply]
Unfortunately, that is not really tenable. If the notification service cannot auth with HA, the URLs must be signed. Otherwise, it could lead to data being leaked via a node traversal /brute forcing. It has nothing to do with the "trustworthiness" of the notification provider. 

For example, the `camera_proxy` URL is `/api/camera_proxy/{entity_id}`, without the signing on the URL (this is designed to replace the token), that URL is _very_ easy to make guesses at until you get a valid URL path.

**Reply by zacwest** [reply]
Right, I mean that we should keep the authentication requirement, just don't pre-sign the URLs. The iOS app can add authentication headers for the requests, as it does now.

**Reply by AngellusMortis** [reply]
Oh, I did actually click your links, so I did not realize you were adding the authentication. In that case then yeah, we would not need it. I think Telegram is capable of passing headers as well for authentication.

**Reply by zacwest** [reply]
If you provide e.g. `/api/camera/proxy/{entity_id}` as an image attachment for the iOS app, it'll add the `Authentication` header when it requests it. This is what it silently does if you provide `data: {entity_id: camera.xyz}` in the service call right now. Any hostname-appropriate path gets this behavior, since it's very common to need it for e.g. media calls.

**Reply by allenporter** [reply]
Given this clarification on how authentication works: what is left in the proposal given this is already all the existing behaviors. Is this just proposing a syntax change for camera entities? Or is the desire to support notification mechanisms that can't do authentication?

Generally, I think it makes sense to make media a "first class citizen" in notifications, but I think the most important reason to do this is for integration into the UI to make building default recipes easier. Right now, you have to go become an expert in the companion app docs to understand the various options (though this has improved a lot in the past 6-12 months!) but the automation builders don't make any of this one click yet (use of blueprints may help, but still media player events use custom events).

From a user perspective, I think what may be missing is glue between an event notification that contains a reference to media and glue to magically generate an attachment for it.  Right now this would require an integration specific blueprint to achieve.  (e.g. when a nest event fires, copy these two fields into a media url and fetch it as a video or image depending on the camera type)

**Reply by AngellusMortis** [reply]
I think it is both things that are important: offering a centralized and seemly way of doing auth for notifications so _both_ users and integration developers have to worry less about it _and_ make it adding images/attachments easier for yours. It is true the auth is already implemented, but it is not currently readily available to users outside of casting.

@balloob's comment above has the perfect example of the user side: https://github.com/home-assistant/architecture/discussions/705#discussioncomment-1951410

**Reply by balloob** [reply]
@allenporter this is for any notification platform, not just mobile app. It could work something like this inside a notify platform of an integration:

```python
to_attach = []
supported = {
  media: ["png", "jpeg", "mp4"],
  sign_urls: False,
}

for attachment in attachments:
	resolved = notify.async_resolve_attachment(attachment, **supported)  
```

Note that in my example https://github.com/home-assistant/architecture/discussions/705#discussioncomment-1951410 I only specified attaching a camera entity without specifying what we would want. I think that we could change the syntax there to offer flexibility.

```yaml
service: notify.balloob
data:
  message: Someone at the door!
  attach:
    - camera:
        entity_id: camera.front_door
        type: snapshot
    - media_source:
        uri: "media-source://local/bla"
```

**Reply by allenporter** [reply]
Yeah, thanks that is helpful,   In my head i imagine the intersection of. [camera integration and/or "types supported": mp4, mpjeg, hls, webrtc, etc], [notification service: ios app, android app, other apps] and today you have to understand the intersection of all of these things to determine what to put in `data`.  So this effectively makes a API for the two systems to negotiate with each other about whats possible

**Comment by dshokouhi** [comment]
One thing to note, the android app does use `entity_picture` attribute for the camera and media player widget to display the image so if we do remove that we would need to supply it in some other method otherwise we would have to change/remove those features.

**Reply by AngellusMortis** [reply]
The idea was not to remove the `camera_proxy` view, but just make it not an attribute. And always authenticated (without a signed URL). That means the app could manually still construct the URL and use it.

But that is a good point if the app uses it, we will definitely need a really long deprecation period. Do we have analytics on the versions of the app that is installed?

**Reply by dshokouhi** [reply]
Looking at Sentry our last release has a 88% adoption rate, when we make this change we will also need to keep it backwards compatible for sometime too.

**Reply by AngellusMortis** [reply]
I do not know what HA upgrade adoption rate looks like for both Web/mobile combined, but my initial idea (based on what I have done at places I have worked) was probably going to be release, wait 6 months to a year, look at analytics, if it is "close enough", deprecate and remove in 1-2 months. 

Even that was met with push back from the mobile devs though because some users will just not upgrade. I would hope with a bunch tech nerds (no offense indented! I am including myself), the adoption rate would be a bit faster.

**Comment by balloob** [comment]
Starting a new answer to focus on implementation details. This is just an initial braindump.

A notify service will be extended with an `attach` option where a user can specify one or more attachments.

It's up to the service to raise an error if it only supports a single attachment or if the combination of the multiple attachments is not allowed.

```yaml
service: notify.balloob
data:
  message: Someone at the door!
  attach:
    - camera.snapshot:
        entity_id: camera.front_door
    - media_source:
        uri: "media-source://local/bla"
```

Any integration can implement one or more attachment resolver. Each attachment contains a single key which specifies the integration and an optional type. The above example has 2 attachments:

- integration `camera`, type `snapshot`
- integration `media_source`, type `None`

It is up to the integration provides the notify service (ie `mobile_app`) to call the `resolve_attachment` helper provided by the notify integration.

The `resolve_attachment` helper will load for each attachment the platform `<integration>/notify.py`. If `ATTACHMENT_RESOLVE_SCHEMA` is defined, it will validate the config. If the config validates, it will call `async_resolve_attachment(hass, type, info, supported)`.

The `info` variable contains the value of the attachment. So it would contain `{ "entity_id": "camera.front_door" }`

The `supported` variable will contain what the notify service supports:
- Audio formats
- Video formats
- Requires URLs or bytes
- Supports authentication or needs HA URLs to be signed
- URL preference (external, internal)

The resolved attachments would be dataclasses. If the notify integration requires bytes, it won't load the bytes but instead offer an iterator for the notify platform to consume (so we can stream data).

**Reply by allenporter** [reply]
I like this as it takes the matrix problem of support and makes it into implementation details that integrations can resolve between themselves. I have less visibility into the broad use cases, but this seems to make sense for the cases I am familiar with.

Thinking through additional implementation details:

- I think `camera` should be able to implemenent `async_resolve_attachment` on behalf of all integrations with camera entities? 
- However, integrations that have media sources (media players?) would need to do this themselves.  (This may be what you already said, but I don't always follow the terminology).
- Seems like `resolve_attachment` can handle the Signing vs Authentication bits, and url preferences.  (`async_resolve_attachment` can give back relative urls).
- One angle we discussed offline is that the `media_source` can be embeded inside an event so that makes it very straight forward to connect a device trigger or event to a notification directly. 
    - Question:  Could it also make sense to attach something to the trigger schema so that an automation builder knows that it is safe to glue these things together? That is, if an event notification already contains a device and a media source, maybe all the glue can be taken care of on behalf of users without a blueprint. (Note: I am not less familiar with how the automation building internals work or how the UI works)

For additional motivation / context to ground this in a concrete example: i'm trying to figure out how we'd document and explain to users how they can configure nest camera with different snapshot types (jpg, mp4, or even mp4 transcoded to gif/jpg?) depending on the camera type, with notifications on android vs ios with different types of media supported.  Then, considering how click through should work through streams would work for different stream types (hls, webrtc).  There is now [literally a table](https://www.home-assistant.io/integrations/nest/#camera) in the integration instructions and it doesn't even have all the possible dimensions yet we're discussing here, and this is just one integration, but an extreme one.

**Reply by AngellusMortis** [reply]
> I think camera should be able to implemenent async_resolve_attachment on behalf of all integrations with camera entities?

That was the idea. Each platform can resolve their own attachments. The notification platform helper would just resolve the platform specific one if it is an entity and then call that one. But that does allow _other_ platforms to create attachments as well (i.e. a `binary_sensor` for motion). But then integrations can extend that further then they add their entities for that platform.

> However, integrations that have media sources (media players?) would need to do this themselves. (This may be what you already said, but I don't always follow the terminology).

Integration _should not_ need custom entrypoints for resolving. You can either make a media_source and resolve it as a normal media source, or extend an entity to resolve it both would resolve with the notify helper.

**Reply by AngellusMortis** [reply]
@balloob 

> integration camera, type snapshot

I am not sure I am quite a fan of this. This may make it harder to extend to new entity types (i.e. a binary_sensor for motion). It would also be hard to do an "auto" mode (i.e. give me the best thing you can that my notification platform supports). I think I was more of a fan of your original idea:

```yaml
entity:
   type: snapshot | stream | clip | auto
   entity_id: camera.front_door
```

`auto` would let us solve for the sending snapshots on Android and HLS streams on iOS. Auto basically could mean stream, fallback to clip, fallback to snapshot, fallback to none. While all of the others are just fallback to none if it cannot be attached.

**Reply by AngellusMortis** [reply]
> Could it also make sense to attach something to the trigger schema so that an automation builder knows that it is safe to glue these things together?

I think it would be good to add the attachment for the device trigger as trigger variable perhaps? So, a device trigger could give an `entity_id` or a media source.

**Reply by balloob** [reply]
The old plan meant the notify attachment resolver needs to know what type maps to what integration. You can still use `camera` instead of `camera.snapshot` and it will also work. Type being `None` could mean try best. 

Integrations implementing attachment resolving could delegate this to the integration if they need to. Camera wouldn’t, it would just talk directly to the entity objects.

Can we stop talking about device triggers please. This thread is about attachment resolving implmenetation.

**Reply by AngellusMortis** [reply]
Sorry you are right. That is off topic and would be a future enhancement after this is implemented. 

And my thought was that it was always just handled by the entity object (if an entity was passed in). So, it would not be a camera entity specific thing. For example, say you want to create a desktop notification (probably HTML5 push) every time the song on your Sonos speakers changed for whatever reason. In that case you would attach your media_player entity:

```yaml
service: notify.balloob
data:
  message: # song name from media player
  attach:
    - entity:
        entity_id: media_player.speaker
        type: snapshot
```

It would be generalized so any entity can provide media if it is capable.

**Reply by balloob** [reply]
Only the integration itself is allowed to interact with it's entity objects. The notify integration is not allowed to talk to a camera entity. So it will need to delegate this to the integration first.

**Reply by AngellusMortis** [reply]
Oh okay! That makes sense then. How would we handle the UI then? Do we have a way to "select" an integration?

Pure YAML would be of no consequence since you can just put whatever you want.

**Reply by balloob** [reply]
We would show a list of eligible entities.

**Reply by allenporter** [reply]
I'd like to talk through next steps of what it would take to get this proposal moving. I regularly see users struggling upstream against the complexity here, and the recent work on automation and media player show that its possible to make significant progress here once we get the right primitives in place.

Assuming we have broad agreement, then I might suggest a PR order such as:
- Implement the `notify` helper for `resolve_attachment` which defines the API for the attachment source side of the integrations, and some of the data classes involved
- Implement an attachment side for `media_source` and the `async_resolve_attachment` details
- Implement a notification side.
    - We could either go straight for the more advanced cases like the mobile apps, or pick something something that has weak attachment support today like `discord` or `telegram` and does not currently support authentication
    - Maybe then we look at mobile notifications which are more advanced at the moment, however they support auth

Alternatively we could start with mobile to work out more of the existing cases.

Given the breadth of the proposal, i'd suggest we implement as needed (e.g. keep `supported` or `type` very minimal) and add capabilities as they are needed by integrations, updating existing integrations as we go.  The proposal sketches out what is possible in the future in detail, but given the large number of features i want to make sure it's aligned with what is actually implemented since this may take a distributed effort.

**Reply by AngellusMortis** [reply]
I think that PR order sounds great. 

Sorry I have not been able to move forward with this at all. I have been really swamped with work the past few weeks. I am basically rebuilding our whole CI/CD infrastructure, so I just have not had the energy to do more coding after works.

Also, I think `notify` targets that _do_ support auth are going to be less common than ones that do. Telegram is the only good example of one that does. I think only one of the major mobile platforms do (iOS or Android), cannot remember which. So, using the signed path as the default is a great way forward.

**Comment by balloob** [comment]
I've moved the signed path part of this discussion to a new issue https://github.com/home-assistant/architecture/discussions/711

**Comment by zach-rosen** [comment]
It looks like the resolution of this was https://github.com/home-assistant/core/pull/64337
Will that be sufficient to bring back thumbnails?

**Comment by AngellusMortis** [comment]
Coming back to this. I am thinking about implementation and usage. I am thinking of the following PR:

1. Add a `has_media` property to the `Entity` class as well as a new `media_urls` attribute (I was thinking a dictionary of `mime type: URL`). These would be signed if necessary and excluded from recorder to prevent JWTs from being stored.
2. Add a `async_resolve_media` method to `Entity` as well as probably a global `async_resolve_media` helper method that any platform can extend to return a playable `MediaItem`
3. Add attach entity functionality to the notify platform

Supplemental PRs:

* The `camera` platform can be updated to support 1 and 3 from above. At a minimum, there would be a URL for the snapshot URL (`camera_proxy`) and maybe also be a way to allow us to finally expose the stream source URLs? This would likely deprecate the `access_token` and `entity_picture` attributes.
* Each of the notify platforms can be updated to support attaching media and automatically use `async_resolve_media` and/or the `media_urls` attribute to generate the necessary URLs for the notify platform.

**Reply by AngellusMortis** [reply]
Thoughts, @balloob and @allenporter?

**Reply by AngellusMortis** [reply]
The idea with `media_urls` is to provide something similar to the [HTML video element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) so an entity can provide all of the media it has and then the consumer (automation, notify platform, etc.) can choose the mime type that are able to use.

**Reply by allenporter** [reply]
Does the camera API need to be extended at all? I think media sources can support arbitrary media already.  The other reason I lean towards this is media for events may be associated with device events and not necessarily be part of the camera entity. (My impression is we suggest this explicitly as a design pattern).

Since this original proposal happened, camera was also extended with a media player which today is used to provide a stream for all cameras. It also has a snapshot (but only in a fallback case, which could be made more explicit with it's own url). I'm not sure that the camera APIs necessarily need to be extended at all given media players can be really flexible: but if there are new additions to media for camera, they could still be exposed as part of the cameras media source anyway).

**Reply by AngellusMortis** [reply]
The issue with media sources is that they require you have access to the entity/python code. There needs to be another generic way to expose media for an entity without access to the Python code for automations, add-ons (Node Red) etc. The recent work and optimization to the recorder has made that really easy: expose them as attributes and then exclude them from recorder. That allows us to combine the way the camera platform exposes `entity_picture` with the new `async_sign_path` to add media an attribute. Additionally, this would allow _any_ entity platform to expose media for an entity (my use case is exposing thumbnail/video clips for Motion/Detections on binary_sensor/sensor entities). 

Camera should only really need to be updated to unify how media is exposed on entities, but I figured this could also be a way to expose streaming URLs which is another thing that I have seen requested as well since there is no way to expose streaming URLs in automations and such at the moment. It could be the groundwork for casting livestreams directly to devices without HA, like the Echo Show without the HLS/LL-HLS proxy (perhaps a direct RTSP proxy if the device supports it).

**Reply by AngellusMortis** [reply]
An example. You have a device for your front doorbell with a camera if a platform decided to expose all of the data:

```
"camera.front_doorbell": {
  "attributes": {
    "media_urls": {
      "rtsp": {
        "video/mpeg": "rtsp://192.168.10.10/url"
      },
      "http": {
        "image/jpeg": "/api/camera_proxy/..."
        "video/mpeg": "/api/camera_proxy_stream/..."
      }
    }
  }
},
"binary_sensor.front_doorbell": {
  "attributes": {
    "media_urls": {
      "http": {
        "image/jpeg": "/api/.../eventthumbnail/...",
        "video/mpeg": "/api/.../eventvideo/..."
      }
    }
  }
}
```

**Comment by allenporter** [comment]
I'd like to revisit this proposal since I think we still see a large need here.

The main focus/simplification of this proposalis: *All attachments are media sources* and we separate attachment resolving concerns from attachment serving concerns.

A notify service will be extended with an attach option where a user can specify one or more attachments.

It's up to the service to decide how to handle supported attachments. A service could also decide to raise if it only supports a single attachment or a combination of attachments is not allowed. Example: mobile app notifications may omit showing attachments if they are not supported or the right format depending on the capabilities of the device (For example, you send image and video and it just shows the image)

This example is the same as the previous example, but with only supporting media sources:
```
service: notify.example_service
data:
  message: Someone at the door!
  attach:
    - media_source:
        uri: "media-source://local/event/thumbnail"
    - media_source:
        uri: "media-source://local/event/video"
```
(This could be simplified further if we want to reduce extensibility)

The previous proposal was surfacing all the requirements and solving them with a single API `resolve_attachment` that returns a dataclass representing the attachment. However, if you look at the existing media source APIs they are good enough to solve most of what we need.

As a refresher, the from the notification implementation perspective the existing APIs is:
```
@dataclass(slots=True)
class PlayMedia:
    """Represents a playable media."""

    url: str
    mime_type: str

async def async_resolve_media(
    hass: HomeAssistant,
    media_content_id: str,
    target_media_player: str | None | UndefinedType = UNDEFINED,
) -> PlayMedia:
    """Get info to play media."""
```

It is up to the integration providing the notify service (ie mobile_app) to call async_resolve_media with the provided uri.

All of the previous requirements around serving content can be satisfied with existing `PlayMedia` API for now:
- Supported formats: The notify service can review the mime type returned by the PlayMedia and decide how best to support the different format based on the mime type.
- Authentication: The notify service may require authentication, or may need HA to sign the URI. It can call `async_sign_path` if needed (from https://github.com/home-assistant/architecture/discussions/711 which was spun off) from the returned PlayMedia uri
- Urls vs Bytes: The PlayMedia returns a URI and the integration can convert that into bytes if it wants to.  If needed, we can create helper libraries for integrations to do this in a standard way.
- URL preference (external, internal) can continue to be handled by the integration itself based on what it needs since PlayMedia uri is a relative path.

Consequences:
- Integrations that want to introduce new attachment types also need to introduce new media source types. (e.g. camera https://github.com/home-assistant/core/pull/65977)
- Integrations may continue to have multiple ways to auth, urls vs bytes, internal vs external, and we could decide to introduce more content serving related APIs in the future to help simplify. (I doubt we would want these to be the same API used to resolve attachments)
- Mime types may not be rich enough, or may not be sufficient, or may be complex to interpret, however we already needed to solve this in the previous proposal. We could introduce additional 
- Integrations may need to implement media sources if they want to provide attachments
- Usability wins given all attachments are media sources and can be used in additional contexts
